src/modules/CRUD/components/Details/Read/AttachmentView/ImportFile/UppySetCategoryBeforeUpload.js

```diff
@@ -0,0 +1,54 @@
+import { Plugin } from '@uppy/core';
+
+class UppySetCategoryBeforeUpload extends Plugin {
+  constructor (uppy, opts) {
+    super(uppy, opts);
+    this.id = this.opts.id || 'SetCategoryBeforeUpload';
+    this.opts = { ...opts };
+    this.type = 'modifier';
+
+    this.setCategoryBeforeUpload = this.setCategoryBeforeUpload.bind(this);
+  }
+
+  getCategory () {
+    const { meta: { category } } = this.uppy.getState();
+    const { props: { findOrCreateAttachmentCategory } } = this.opts;
+    return findOrCreateAttachmentCategory(category);
+  }
+
+  async setCategoryBeforeUpload (fileIDs) {
+    const promises = fileIDs.map(fileID => {
+      const file = this.uppy.getFile(fileID);
+      this.uppy.emit('preprocess-progress', file, {
+        mode: 'indeterminate',
+      });
+      return fileID;
+    });
+
+    const emitPreprocessCompleteForAll = async () => {
+      const category = await this.getCategory();
+      fileIDs.forEach(fileID => {
+        const file = this.uppy.getFile(fileID);
+        file.meta.category = category.id;
+        this.uppy.emit('preprocess-complete', file);
+      });
+    };
+
+    // Why emit `preprocess-complete` for all files at once, instead of
+    // above when each is processed?
+    // Because it leads to StatusBar showing a weird “upload 6 files” button,
+    // while waiting for all the files to complete pre-processing.
+    return Promise.all(promises)
+      .then(emitPreprocessCompleteForAll);
+  }
+
+  install () {
+    this.uppy.addPreProcessor(this.setCategoryBeforeUpload);
+  }
+
+  uninstall () {
+    this.uppy.removePreProcessor(this.setCategoryBeforeUpload);
+  }
+}
+
+export default UppySetCategoryBeforeUpload;
```

Comment:
As this seems to be a plugin for Uppy, and has nothing in common with a component (nor a service dedicated to a component), shouldn't this be put appart in a dedicated `lib`-like directory, outside of `/components` directory?

Reply:
Also, couldn't this be contributed as a dedicated plain Uppy plugin?

Reply:
In fact, it's very specific for the usecase. For a better understanding, below my approach ⬇️ 

My goal was to set category to files before uploading them.

The endpoint to POST a new attachment has a required category ID. To be able to submit files, the user must choose a category. Once selected, and then clicked on the upload button, there are two cases:
* The category already exists so we set its related ID.
* The category does not exist and we POST the new category and await the response of the call to set the new ID.

 `uppy` has a method `onBeforeUpload` to do this kind of manipulation, but it does not wait for asynchronous function https://uppy.io/docs/uppy/#onBeforeUpload-files-gt-files

>Note that this method is intended for quick synchronous checks/modifications only. If you need to do an async API call, or heavy work on a file (like compression or encryption), you should utilize a custom plugin instead.

That's why i built this plugin.

Reply:
Good explanation, it still seems like as a good plugin to spread to the community!
This plugin add a feature that might help a lot of other developers.

Maybe not exactly as is: You could take the `setCategoryBeforeUpload` function out of the plugin and provide it as a plugin argument.
Then, the main intent of the plugin would be to allow awaiting an `onBeforeUpload` call.

Doing it this way is not mandatory, but it is clearly a proper business & logic opportunity.